aten/src/ATen/native/cuda/Distributions.cu

```diff
@@ -69,7 +69,13 @@ struct curand_uniform_wrapper {
   curandStatePhilox4_32_10_t &state;
   __device__ curand_uniform_wrapper(curandStatePhilox4_32_10_t &state): state(state) {}
   __device__ float operator()() {
-    return curand_uniform(&state);
+    auto val = curand_uniform(&state);
+    // val is from 0.0 to 1.0, where 1.0 is included and 0.0 is excluded
+    // We want 1.0 excluded and 0.0 included
+    if (val == 1.0F) {
+        val = 0.0F;
```

Comment:
now you are oversampling 0s, severely, due to some curand quirks (curand divides uint by 2^32, so what would happen here is that 0 will be generated with the probability of 2^-24, because that's how often 1 is generated, but then 2^-32 will be generated with the probability 2^-32. The way to make it consistent with the rest of pytorch is to use transformation similar to the one in TransformationHelper that I linked (which generates 0, 2^-24, 2^-23 .... 1-2^24  with the same probability).

Reply:
I would have thought that kind of mistake would have been caught by tests. Is this a known problem with `curand_uniform()`? In any case I adopted the code you linked to.